src/inject/dynamic-theme/css-rules.ts

```diff
@@ -41,23 +42,33 @@ function isCSSVariable(property: string) {
     return property.startsWith('--') && !property.startsWith('--darkreader');
 }
 
+function getParentGroups(rule: CSSRule, stack: string[] = []): string[] {
+    const {parentRule} = rule;
+    if (!parentRule || !(parentRule instanceof CSSMediaRule)) {
+        return stack;
+    }
+    stack.push(parentRule.conditionText);
+    return getParentGroups(parentRule, stack);
+}
+
 export function getCSSVariables(rules: CSSRuleList) {
-    const variables = new Map<string, string>();
+    const variables = new Map<string, Variable>();
     rules && iterateCSSRules(rules, (rule) => {
+        const parentGroups = getParentGroups(rule);
         rule.style && iterateCSSDeclarations(rule.style, (property, value) => {
             if (isCSSVariable(property)) {
-                variables.set(property, value);
+                variables.set(`${rule.selectorText};${property}`, {property, value, parentGroups});
             }
         });
     });
     return variables;
 }
 
 export function getElementCSSVariables(element: HTMLElement) {
-    const variables = new Map<string, string>();
+    const variables = new Map<string, Variable>();
     iterateCSSDeclarations(element.style, (property, value) => {
         if (isCSSVariable(property)) {
-            variables.set(property, value);
+            variables.set(`:root;${property}`, {property, value, parentGroups: []});
```

Comment:
Hmmm, since it will be used for `<html>` element only at the moment, let's use `:root`, but later something other should be invented (e.g. assigning some `data-` attribute to make it a selector, like it is currently made for overriding inline colors).